source/endpoints_regex.c

```diff
@@ -0,0 +1,312 @@
+        aws_array_list_get_at_ptr(symbols, (void **)&element, i);
+
+        if (element->type == AWS_ENDPOINTS_REGEX_SYMBOL_ALTERNATION_GROUP) {
+            aws_string_destroy(element->info.alternation);
+        }
+    }
+}
+
+struct aws_endpoint_regex *aws_endpoint_regex_new_from_string(
+    struct aws_allocator *allocator,
+    struct aws_byte_cursor regex_pattern) {
+
+    if (regex_pattern.len == 0 || regex_pattern.len > s_max_regex_length) {
+        AWS_LOGF_ERROR(
+            AWS_LS_SDKUTILS_ENDPOINTS_REGEX,
+            "Invalid regex pattern size. Must be between 1 and %zu",
+            s_max_regex_length);
+        aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+        return NULL;
+    }
+
+    if (regex_pattern.ptr[0] != '^' || regex_pattern.ptr[regex_pattern.len - 1] != '$') {
+        AWS_LOGF_ERROR(
+            AWS_LS_SDKUTILS_ENDPOINTS_REGEX,
+            "Unsupported regex pattern. Supported patterns must match the whole text.");
+        aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_UNSUPPORTED_REGEX);
+        return NULL;
+    }
+
+    /* Ignore begin/end chars */
+    aws_byte_cursor_advance(&regex_pattern, 1);
+    --regex_pattern.len;
+
+    struct aws_array_list *symbols = aws_mem_calloc(allocator, 1, sizeof(struct aws_array_list));
+    aws_array_list_init_dynamic(symbols, allocator, regex_pattern.len, sizeof(struct aws_endpoint_regex_symbol));
+
+    while (regex_pattern.len > 0) {
+        uint8_t ch = regex_pattern.ptr[0];
+        aws_byte_cursor_advance(&regex_pattern, 1);
+
+        struct aws_endpoint_regex_symbol symbol;
+        switch (ch) {
+            case '.':
+                symbol.type = AWS_ENDPOINTS_REGEX_SYMBOL_DOT;
+                break;
+            case '*':
+                symbol.type = AWS_ENDPOINTS_REGEX_SYMBOL_STAR;
+                break;
+            case '+':
+                symbol.type = AWS_ENDPOINTS_REGEX_SYMBOL_PLUS;
+                break;
+            case '\\':
+                switch (regex_pattern.ptr[0]) {
+                    /* Predefined patterns */
+                    case 'd':
+                        symbol.type = AWS_ENDPOINTS_REGEX_SYMBOL_DIGIT;
+                        break;
+                    case 'w':
+                        symbol.type = AWS_ENDPOINTS_REGEX_SYMBOL_ALPHA;
+                        break;
+
+                    /* Escaped chars, ex. * or + */
+                    default:
+                        symbol.type = AWS_ENDPOINTS_REGEX_SYMBOL_CHAR;
+                        symbol.info.ch = regex_pattern.ptr[0];
+                        break;
+                }
+                aws_byte_cursor_advance(&regex_pattern, 1);
+                break;
+            case '(': {
+                struct aws_byte_cursor group = {0};
+                if (!aws_byte_cursor_next_split(&regex_pattern, ')', &group)) {
+                    AWS_LOGF_ERROR(
+                        AWS_LS_SDKUTILS_ENDPOINTS_REGEX, "Invalid regex pattern. Missing closing parenthesis.");
```

Comment:
This log statement seems incorrect since `aws_byte_cursor_next_split` will match even if there was no `)` parenthesis.